File: markov.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import webapp2
import os
from google.appengine.ext.webapp import template
from xml.etree.ElementTree import *
import re

import urllib

class Markov(webapp2.RequestHandler):
	def get(self):
		mes=""
		mes=self.get_hotentry_title()
		
		template_values={
		'mes':mes
		}
		path = os.path.join(os.path.dirname(__file__), 'html/markov.html')
		self.response.out.write(template.render(path, template_values))
		
	def get_hotentry_title(self):
		titles = ""
		tree = parse(urllib.urlopen('http://feeds.feedburner.com/hatena/b/hotentry'))
		for i in tree.findall('./{http://purl.org/rss/1.0/}item'):
			titles+= re.sub("[-:|/｜：].{1,30}$","",i.find('{http://purl.org/rss/1.0/}title').text) + "\n"
		return titles
		
	def get_2ch(self):
		titles = ""
		response = urllib.urlopen('https://leia.5ch.net/poverty/subject.txt')
		html = unicode(response.read(), "cp932", 'ignore').encode("utf-8")
		for line in html.split("\n"):
			if line != "":
				titles+=re.sub("\(.*?\)$","",line.split("<>", 2)[1])+ "\n"
		return titles
		
app = webapp2.WSGIApplication([
	('/markov.html', Markov)
], debug=True)

File: markov.go
package main

import (
	"encoding/xml"
	"fmt"
	"html/template"
	"io/ioutil"
	"log"
	"net/http"
	"regexp"
)

func MarkovHandler(w http.ResponseWriter, r *http.Request) {
	t, err := template.ParseFiles("html/markov.html")
	if err != nil {
		log.Fatalf("template error: %v", err)
	}
	if err := t.Execute(w, struct {
		Message string
	}{
		Message: getHotentry(),
	}); err != nil {
		log.Printf("failed to execute template: %v", err)
	}
}

type RSS struct {
	XMLName xml.Name `xml:"RDF"`
	Items   []Item   `xml:"item"`
}

type Item struct {
	XMLName xml.Name `xml:"item"`
	Title   string   `xml:"title"`
}

func getHotentry() string {
	url := "http://feeds.feedburner.com/hatena/b/hotentry"
	resp, err := http.Get(url)
	if err != nil {
		fmt.Println(err.Error())
		return ""
	}
	defer resp.Body.Close()
	byteArray, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		fmt.Println(err.Error())
		return ""
	}
	v := RSS{}
	err = xml.Unmarshal(byteArray, &v)
	if err != nil {
		fmt.Println(err.Error())
		return ""
	}
	s := ""
	rep := regexp.MustCompile(`[-:|/｜：]`)
	for i := 0; i < len(v.Items); i++ {
		result := rep.Split(v.Items[i].Title, -1)
		if len(result) > 0 {
			s += result[0] + "\n"
		}
	}
	return s
}
